plaso/parsers/postgresql.py

```diff
@@ -0,0 +1,184 @@
+# -*- coding: utf-8 -*-
+"""Parser for PostgreSQL log files."""
+
+import string
+import pyparsing
+
+from dateutil import parser
+
+from dfdatetime import time_elements as dfdatetime_time_elements
+
+from plaso.containers import events
+from plaso.containers import time_events
+from plaso.lib import definitions, errors
+from plaso.parsers import logger
+from plaso.parsers import manager
+from plaso.parsers import text_parser
+
+class PostgreSQLEventData(events.EventData):
+  """PostgreSQL log data.
+
+  Attributes:
+    log_level (str): logging level of event
+    log_line (str): log message.
+    pid (int): process identifier (PID).
+    user (str): "user@database" string if present.
+  """
+
+  DATA_TYPE = 'postgresql'
+
+  def __init__(self):
+    """Initializes event data."""
+    super(PostgreSQLEventData, self).__init__(data_type=self.DATA_TYPE)
+    self.log_level = None
+    self.log_line = None
+    self.pid = None
+    self.user = None
+
+class PostgreSQLParser(text_parser.PyparsingMultiLineTextParser):
+  """Parses events from PostgreSQL log files."""
+
+  NAME = 'postgresql'
+  DATA_FORMAT = 'PostgreSQL log file'
+  _ENCODING = 'utf-8'
+
+  _DATE_TIME = pyparsing.Group(
+      pyparsing.LineStart() +
+      text_parser.PyparsingConstants.DATE_ELEMENTS +
+      text_parser.PyparsingConstants.TIME.setResultsName('time') +
+      pyparsing.Optional(
+          pyparsing.Suppress('.') +
+          text_parser.PyparsingConstants.INTEGER
+      ).setResultsName('microseconds') +
+      pyparsing.Word(pyparsing.printables).setResultsName('time_zone')
+  ).setResultsName('date_time')
+
+  _PID = pyparsing.Group(
+      pyparsing.Suppress('[') +
+      pyparsing.OneOrMore(text_parser.PyparsingConstants.INTEGER) +
+      pyparsing.Optional(pyparsing.Literal('-')) +
+      pyparsing.ZeroOrMore(text_parser.PyparsingConstants.INTEGER) +
+      pyparsing.Suppress(']')
+  ).setResultsName('pid')
+
+  _USER_AND_DATABASE = pyparsing.Group(
+      pyparsing.Word(pyparsing.alphanums) +
+      pyparsing.Literal('@') +
+      pyparsing.Word(pyparsing.alphanums)
+  ).setResultsName('user_and_database')
+
+  _LOG_LEVEL = (
+      pyparsing.Word(string.ascii_uppercase) +
+      pyparsing.Suppress(':')
+  ).setResultsName('log_level')
+
+  _LOG_LINE_END = _BODY_END = pyparsing.StringEnd() | _DATE_TIME
+
+  _LOG_LINE = (
+      _DATE_TIME +
+      _PID +
+      pyparsing.Optional(_USER_AND_DATABASE) +
+      _LOG_LEVEL +
+      pyparsing.SkipTo(_LOG_LINE_END).setResultsName('log_line') +
+      pyparsing.ZeroOrMore(pyparsing.lineEnd())
+  )
+
+  LINE_STRUCTURES = [
+    ('logline', _LOG_LINE),
```

Comment:
style nit: extra 2 spaces.  alternatively can this fit on a single line?

Reply:
Done